src/iceberg/avro/avro_direct_decoder.cc

```diff
@@ -0,0 +1,597 @@
+      if (avro_node->type() != ::avro::AVRO_INT) {
+        return InvalidArgument("Expected Avro int for int field, got: {}",
+                               ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::Int32Builder*>(array_builder);
+      int32_t value = decoder.decodeInt();
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(value));
+      return {};
+    }
+
+    case TypeId::kLong: {
+      auto* builder = internal::checked_cast<::arrow::Int64Builder*>(array_builder);
+      if (avro_node->type() == ::avro::AVRO_LONG) {
+        int64_t value = decoder.decodeLong();
+        ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(value));
+      } else if (avro_node->type() == ::avro::AVRO_INT) {
+        int32_t value = decoder.decodeInt();
+        ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(static_cast<int64_t>(value)));
+      } else {
+        return InvalidArgument("Expected Avro int/long for long field, got: {}",
+                               ToString(avro_node));
+      }
+      return {};
+    }
+
+    case TypeId::kFloat: {
+      if (avro_node->type() != ::avro::AVRO_FLOAT) {
+        return InvalidArgument("Expected Avro float for float field, got: {}",
+                               ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::FloatBuilder*>(array_builder);
+      float value = decoder.decodeFloat();
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(value));
+      return {};
+    }
+
+    case TypeId::kDouble: {
+      auto* builder = internal::checked_cast<::arrow::DoubleBuilder*>(array_builder);
+      if (avro_node->type() == ::avro::AVRO_DOUBLE) {
+        double value = decoder.decodeDouble();
+        ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(value));
+      } else if (avro_node->type() == ::avro::AVRO_FLOAT) {
+        float value = decoder.decodeFloat();
+        ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(static_cast<double>(value)));
+      } else {
+        return InvalidArgument("Expected Avro float/double for double field, got: {}",
+                               ToString(avro_node));
+      }
+      return {};
+    }
+
+    case TypeId::kString: {
+      if (avro_node->type() != ::avro::AVRO_STRING) {
+        return InvalidArgument("Expected Avro string for string field, got: {}",
+                               ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::StringBuilder*>(array_builder);
+      std::string value;
+      decoder.decodeString(value);
+      ICEBERG_ARROW_RETURN_NOT_OK(builder->Append(value));
+      return {};
+    }
+
+    case TypeId::kBinary: {
+      if (avro_node->type() != ::avro::AVRO_BYTES) {
+        return InvalidArgument("Expected Avro bytes for binary field, got: {}",
+                               ToString(avro_node));
+      }
+      auto* builder = internal::checked_cast<::arrow::BinaryBuilder*>(array_builder);
+      std::vector<uint8_t> bytes;
```

Comment:
Do we want to avoid frequent small object allocation like this? Perhaps we can reuse it by adding a `class DecodeContext` where a `std::vector<uint8_t> scratch` object is supposed to be used from it.

Reply:
Same for other temp vector and string variables.